### ISSUE TITLE
Revisiting versioning strategy of the JSON API spec

### ISSUE THREAD

**Issue by hhware** [json-api/json-api#432]
**(1)** AFAIK, versioning strategy of JSON API spec is intentionally left [undefined](https://github.com/json-api/json-api/issues/46#issuecomment-17656083). IMHO, since the spec does have versions (1.0, RC2, etc. are being widely referenced), it would be useful to specify how these labels are constructed.

**Proposal:** add _"JSON API is versioned according to the specification of [Semantic Versioning](http://semver.org/)"_ somewhere in the Introduction. If SemVer is adopted, 1.0 should obviously become 1.0.0.

**(2)** Even though JSON API will not be introducing backward-incompatible changes after 1.0, nothing prevents it from introducing significant changes in a backward-compatible manner, possibly multiple times over the years. When a client-side library implementing JSON API 1.N is processing a response from a server implementing JSON API 1.M, where N is much higher than M, that client should be able to guess availability of features available on the server from the response format. However, it could be argued that knowing M first-hand from the server could simplify the job of the client.

In addition to that, if it so happens that JSON API actually decides for some reason to introduce backward-incompatible changes in the future, despite the current intentions not to ever do that, transition will be much easier for clients if version of the format is always explicitly communicated. IMHO, it does not hurt to be cautious, especially if the only expense in this case is to add a few characters to headers...

**Proposal**: require that both client and server advertise JSON API version they are using via media type parameter, for instance:

```
Content-Type: application/vnd.api+json; json-api-version=1.0.0
```

**(3)** If (2) is accepted, it could be immediately useful for those who have already started using JSON API: the draft was out in the wild for almost two years! For example, the project which I am involved in uses pre-RC2 JSON API spec and advertises `Content-Type: application/vnd.api+json`. Since RC2, our system is formally incompatible with JSON API, but we cannot even change content type now because of the existing integrations with clients, etc. This would quickly solve this petty problem of ours:

```
Content-Type: application/vnd.api+json; json-api-version=1.0.0-rc.1
```

I would be glad to submit a PR for any/all of this.

**Comment by dgeb** [comment]
Marked for discussion before 1.0.

**Comment by eneuhauser** [comment]
It would be helpful to have this now and add details to the "Update History". After implementing a server-side implementation using the "RC2" standard, I learned the [ember-json-api](https://github.com/kurko/ember-json-api) client I was hoping to use was behind. I submitted a [pull request](https://github.com/kurko/ember-json-api/pull/68) to make it compliant with the latest version, but now my 9 day old pull request is out-of-date.

As other developers want to consume existing libraries, there should be common semantics around what version of JSON API that library supports.

**Comment by dgeb** [comment]
We are truly committed to maintaining backwards compatibility once 1.0 stable is released. Only additive changes will be allowed from that point forward.

This has been the plan from the very beginning (see #46). And we're almost completely done with last minute changes before 1.0 stable.

**Comment by steveklabnik** [comment]
Yes, this is something I feel incredibly strongly about, but, with SemVer, it means that pre 1.0.0, you can do whatever you want. That's the spirit here: spend two years (!) making changes, to feel out the space, and then nail down a 1.0 when it feels comfortable, and then never make them again.

**Comment by ColtonProvias** [comment]
@eneuhauser That's the issue I feel with my sqlalchemy-jsonapi project.  My previous incarnation of it (here on GitHub) is very out of date and right now I'm pulling it again from some private code on my own private git server and trying to get it up to spec before replacing the GitHub one with the new one.  However, there are some areas where it's still problematic for me to implement (relationships, pagination, filtering).  So I'm trying to help get this spec cleaned up so I can get that library released and to spec, along with getting my own projects that depend on JSON-API completed.

**Comment by gr0uch** [comment]
I opened this exact issue recently and @steveklabnik closed it immediately.  it seems that churn is a huge issue and why I don't want to consider json api until its 1.0. moreover, most of its MAYs are not advertised by the server which is what keeps changing the most, and there hasn't been much discussion on how to do feature detection. for example, there is no way to know if a server supports inclusions unless making a request with the include query and missing the `included` key. the contracts that it provides seems shaky, possibly even after 1.0.

**Comment by ColtonProvias** [comment]
@daliwali Wouldn't it be possible to include feature detection as an extension?  Let's say that you have the following URL pattern:

`/<type>/<id>/`

Then maybe you could "describe" it and other endpoints for that type via something like:

`/describe/<type>/`

It could then deliver a response such as:

``` json
{
  "data": {
    "endpoints": [
      {
        "method": "GET",
        "url": "https://mysite.com/api/posts/{id}/",
        "extra_fields": [],
        "excluded_fields": [],
        "include": true,
        "filter": false,
        "page": false,
        "sort": true,
        "fields": true,
        "meta": {
          "docs": "https://mysite.com/api_docs/posts/id/GET/"
        }
      }
    ],
    "resource": {
      "type": "posts",
      "fields": [
        {
          "name": "title",
          "type": "str",
          "validators": [
            "required",
            {
              "type": "length",
              "min": "5",
              "max": null
            }
          ]
        }
      ]
    }
  }
}
```

**Comment by gr0uch** [comment]
@ColtonProvias I think the point is that by not describing feature detection as part of the _base_ specification, leaving it optional, then it remains ambiguous to clients what JSON API 1.0 means. When features may be added post-1.0, it will remain unclear what 1.0 means if they may not be assumed to be implemented. 

I think that this is a very ambitious spec that overextends into the mechanical aspects of HTTP and imposing many features on the server, and the base specification prescribes many features that may be independent specifications on their own (case in point: JSON Patch, which was dropped). Despite it being rather comprehensive, the spec leaves much open to interpretation and ambiguity.

**Comment by dgeb** [comment]
See #449 - this formalizes the ability of this specification to evolve through additive changes. Once at 1.0, it will not break backward compatibility.

**Comment by taion** [comment]
Strictly from a convenience perspective, it would help to have version numbers to talk about supported features. From my perspective as a user, it'd be easier for me to pick up a library that claimed to support some set of JSON API features up to version x.y.z and understand what I can and cannot do, as opposed to having to manually process each of a potentially long list of possible features - e.g. if a server implementation says it supports all optional functionality in the core format as of version 1.1.5 or something, that's actually fairly valuable to me as shorthand for understand what features are or are not available.

**Comment by dgeb** [comment]
Any attempt to advertise a version after 1.0 lands seems counter to the promise of backward compatibility. What does seem important is feature discovery for a server in general and at specific endpoints, as discussed in #304. But I'm not sure that needs to be solved by 1.0. We could define an extension to provide that facility with separate requests.

**Comment by taion** [comment]
As long as the version remains 1.x.x, I as a user would absolutely expect backward compatibility to fully be maintained. In this case, you are just asserting that there will never be a 2.x.x; this is fine, but, separate from the mechanical discoverability issue, I care about figuring out whether e.g. a server library has full support for optional features added in 1.4 v. only having support for features up to 1.3, say. My concern here is much more about understandability for human consumers of libraries in having a good way to tell what features those libraries can be expected to have.

**Comment by taion** [comment]
To be more concrete, the thing I'm talking about is not figuring out as a JSON API client whether a server supports some feature I want. It's figuring out as a developer whether e.g. https://github.com/coltonprovias/sqlalchemy-jsonapi or another library implements support for such a feature.

**Comment by ColtonProvias** [comment]
Would semantic versioning really be the best solution?  In the end, the client is just going to compare it to an acceptable range and isn't going to care about how we humans see it.  We can do it by sequential numbering, by semantic versioning, by approaching pi, by unix timestamp, by entropy, or any other means but in the end only the client has to actually deal with this number.  The way a version is represented to humans is, essentially, just a cosmetic thing.  I say we target achieving 1.0.0 first, get the rest of the problem areas taken care of, and then we should pick this discussion back up once we need to release the first update.

As to declaring an implementation complete, partial, or extended (in reference to @taion), it must be noted explicitly by the developer of the library or in the API docs of a site what the project's compatibility with JSON-API is.  For example, I just updated SQLAlchemy-JSONAPI that @taion referenced to state that it is not compatible with the current spec.  Once I get it fully compatible, I will explicitly state full compatibility.  And if I implement, let's say the bulk extension or custom pagination and filter backends, those will also be noted explicitly so developers are aware.

**Comment by taion** [comment]
Fair enough. Where the lack of versioning has been annoying to me has actually been in terms of figuring out whether the current libraries support RC2 or not - if the pre-release versions had version numbers, maybe the libraries could have stated what version of JSON API they supported. This sort of complete breaking change is a non-issue after hitting release, of course.

My counter-argument, though, would be that, if, say, an amazing new feature were to be added in the first release after 1.0.0 (let's say 1.1.0), if you don't identify the initial release as v1.0.0, I wouldn't be able to e.g. look at SQLAlchemy-JSONAPI, see that it says it only supports version 1.0.0, and conclude that the new version 1.1.0 features would not be available there.

Consider the case of programming languages. C++11/14/17 are (mostly) backward compatible with older versions, but it's convenient to have a version number to refer to, rather than just thinking of them as "C++ with a bag of extra features". Or likewise ES6/ES7... at least I get annoyed every time I have to think about what specific ES6 features my transpiler actually supports.

**Comment by ColtonProvias** [comment]
Of course there will be a version number and it would refer to what is supported.  It's just that right now the API is changing so drastically that assigning a version number to it at this state isn't necessarily useful.  We went from PUT one day to PATCH the next (which irked me because I just got PUT implemented in the new SQLAlchemy-JSONAPI that day just after moving PATCH to an extension).

Until 1.0.0 is released, it should be assumed that not all features of this spec will be implemented and that JSON-API is not yet production ready.

Once 1.0.0 is released, then it will give us library developers a target to implement.  Until then, don't get too hasty.  If you want to get this spec finalized as well, then start helping out in the discussions and get it closer to completion!

Sincerely,
Someone who was too hasty and implemented this spec in 3 projects pre-maturely.  It's a mess now on my hands.

**Comment by hhware** [comment]
> Any attempt to advertise a version after 1.0 lands seems counter to the promise of backward compatibility. 

@dgeb, sorry, I am totally missing, why is that so? E.g., if SemVer is adopted, then the fact that the version is set to 1.x.y _asserts_ that it is backward compatible with 1.0.0, which seems to be completely opposite to your statement. If none of existing policies is adopted, then one needs to define a custom policy to give meaning to version numbers. But I personally find it hard to see how being intentionally obscure about the spec version could be related to backward compatibility or lack thereof.

> What does seem important is feature discovery for a server in general and at specific endpoints, as discussed in #304.

Isn't #304 mostly about a specific case of feature discovery, namely, about verbs on endpoints? Let's consider another possible feature: a new optional attribute is added to top-level `links`. Since it is optional, it does not seem easy for the client library to tell from a given response whether it was skipped in that particular response or the server is running an older version of JSON API, which does not have this attribute declared. If spec version is advertised, it is a non-issue.

> Would semantic versioning really be the best solution?

@ColtonProvias, I think SemVer is a good fit because it is very short (the whole thing is just 11 items) and it is built around the notion of backward compatibility, which is very dear for the spec authors. If there are better options, fine, let's adopt them...

> Once I get it fully compatible, I will explicitly state full compatibility.

Exactly! But why not just state it in the form of 

```
Content-Type: application/vnd.api+json; json-api-version=1.0.0
```

Then it is not only stated in the library docs, but can be used by both humans and peer libraries, if need be...

Sorry for being annoying, but are there any specific arguments why versioning policy _should not_ be defined and why spec version _should not_ be communicated explicitly in the request/response? Most of the discussion in this ticket seems to be about backward compatibility, which I believe is an orthogonal question to the versioning policy: IMHO, it is totally fine to be backward compatible, to define versioning policy, and to require version to be explicitly communicated _at the same time_...

**Comment by ethanresnick** [comment]
I agree that, for helping developers understand a library's feature support etc., JSON API should have a version number associated with its changes, even if those changes are backwards compatible. And I don't think this should be controversial. E.g. Mark Nottingham, who's [written](https://www.mnot.net/blog/2011/10/25/web_api_versioning_smackdown) a lot about how versioning APIs is not analogous to versioning desktop software, [says](https://www.mnot.net/blog/2012/12/04/api-evolution):

> There are legitimate needs for minor and patch information; just because we say "don't put them into names" doesn't mean that they shouldn't exist... So, where should the minor and patch version numbers go? Easy -- it's useful for the release notes, and a few other forms of documentation. It's useful for marketing. In the case of a more complex API (as with most standards -- whether they come out of a standards body or an open source project), it's useful for packaging up an agreed-to set of functionality and calling that a spec release.

So, I think the question is just how to _communicate_ the version over the wire, if at all.

Communicating it in a media-type parameter doesn't seem  _that_ bad to me, but I could be overlooking some of negative side-effects. Right now, the only problem I see is that, if stupid clients put the version parameter into the `Accept` header, it could limit the cacheability of responses if the server is supporting content-negotiation.

For example, imagine an API has two clients, one of which sends a request with the header `Accept: application/vnd.api+json; version=1.1.1` and the other sends `Accept: application/vnd.api+json; version=1.1.4`. Regardless of what version the server knows how to generate, it can respond with a document of that version (possibly excluding some newer fields to save bandwidth), because it knows that all json-api changes are backwards-compatible. That's key, because we'd otherwise have to specify what to return when there's no version provided. So, really, all the version parameter would be doing is broadcasting server support for certain features.

Now, the problem comes in if the server also includes a `Vary: Accept` header in its responses. That header might be there, e.g., so that a web browser can request `/people/1` and see an HTML page for the first person while a JSON-API client can request the same URI with an `Accept: application/vnd.api+json` and get the JSON-API response. With that header there, suddenly a response to the first client that requested `version=1.1.1` can't be returned by a cache to the second client, which specifies a different `Accept` header.

So, what to do about this? If it's really the only issue with using a media type parameter (which I kinda doubt), then we could probably just make a note telling clients not to put a specific version in their `Accept` headers. (Especially considering that `Accept` headers will [vary a lot](https://wiki.whatwg.org/wiki/Why_not_conneg#Negotiating_by_format), even if we don't have a version parameter.) That said, we can always add a parameter to the media-type post 1.0, and I think there should be a compelling use case before we do so.

**Comment by ethanresnick** [comment]
One concrete case: sideposting. We're probably going to add it at some point, and adding it probably won't require the creation of new endpoints. Which means that a server won't be able to advertise support for sideposting through conventional means, like `Allow` and/or JSON Home. So how should a client know whether to attempt a sidepost or make a series of separate POSTs? Imo, a media type version parameter that the client can capture from a previous response's `Content-Type` doesn't seem like a horrible solution.

**Edit**: from this POV, the `version` parameter could be seen as just a numeric equivalent of `supported-ext`, but used to describe supported, optional features from the base spec, as opposed to supported, optional extensions. By that logic, we could also just call this parameter `supported-features`, where its value would be something like: `include, fields`. And, actually, that makes more sense, as it's possible to imagine a library that supports some features from 1.1.0 while still not supporting all the optional features from 1.0. 

So, my vote is now: have versions for documentation's sake and easy reference (e.g. a developer can say "My library supports all optional features from the 1.2 spec") but communicate over the wire only a string of supported, optional features, rather than a numeric version.

**Comment by taion** [comment]
@ColtonProvias
But this issue is on the 1.0 milestone! :smile_cat:  

@ethanresnick 
I agree with your proposal to enumerate features on the wire, but to use version numbers for documentation. Just having version numbers on the wire would not be sufficient to deal with all the optional and extension bits, but as a developer I'd strongly prefer just to have a version number as shorthand when talking about features.

**Comment by taion** [comment]
One way to make the version information really structured and machine-parseable is just to get rid of all the optional parts of the base specification, to state them only in terms of **SHOULD** and reserved words, and to move that functionality into extensions (which then have their own semantic versioning).

One example here would be the optional sorting spec - keeping it as an optional part of the core spec would require figuring out a way to communicate support for it separate from support for extensions, but moving it out of the core spec would let you use the same mechanism as for other extensions.

**Comment by tgwizard** [comment]
Couldn't feature detection be communicated out-of-band (e.g. documentation), and not be part of the JSON-API spec at all? Or is it the idea that clients should be able to automatically/programmatically detect which features are available? It seems to me as if the scope for JSON-API is growing by the day :disappointed:

**Comment by dgeb** [comment]
> One way to make the version information really structured and machine-parseable is just to get rid of all the optional parts of the base specification, to state them only in terms of SHOULD and reserved words, and to move that functionality into extensions (which then have their own semantic versioning).

Extensions could be created that are additive to the spec and convert a MAY in the base spec into a MUST. For instance, a hypermedia extension might require `self` and `related` links.

I don't see a problem leaving MAYs in the base spec however.

> Couldn't feature detection be communicated out-of-band (e.g. documentation)

Yes, that's definitely one path. 

> communicate over the wire only a string of supported, optional features

A separate extension could, and probably should, be created for feature detection. Not everything needs to be in the base spec. I suspect this will take some time to evolve properly.

> have versions for documentation's sake and easy reference

This seems fine to include with the spec's documentation to reflect that the spec will evolve additively. I'm fine with tagging versions of the spec and displaying a history of changes since 1.0.

However, I think that a media type parameter specifying a `json-api-version` is too rigid because it implies that media type negotiation will involve the version. It also implies that extensions themselves should be versioned, which leads to an n-dimensional grid of complexity.

**Comment by taion** [comment]
The reason I bring up extensions is that it seems like they already get you most of the way toward feature detection. There's something neat about not needing something separate to specify e.g. whether a given endpoint supports sorting as described in the base spec, and just saying that it supports e.g. the `json-api/sort-v1.0` extension or something.

I just think it's tidier in some sense to consistently communicate concrete support for optional features in broadly understandable way with the current extension system. It's just an economy of concepts thing - i.e. saying that I support the base spec + some set of extensions, rather than saying that I support the base spec + the optional sorting spec from the base + some set of extensions.

(FWIW, I actually don't really care about automated feature detection.)

**Comment by hhware** [comment]
@ethanresnick, thanks for the informative reference!

> we could also just call this parameter `supported-features`, where its value would be something like: `include, fields`

It is a possibility, but I think it will require a lot of effort to maintain the spec accurately in order to minimize possibility of inadvertent change to one feature while changing another. Also, some features may depend on others. It seems to be pretty complex, is this effort justifiable to accommodate implementers who wish to ignore part of the spec? The goal of this propsal was much more simple and practical...

> So, really, all the version parameter would be doing is broadcasting server support for certain features.

Exactly! What to do with it, is left for the peer to decide, but at least it _enables_ the peer to make an informed decision.

> A separate extension could, and probably should, be created for feature detection. Not everything needs to be in the base spec. I suspect this will take some time to evolve properly.

I totally agree with that. However, even if feature detection ends up implemented not via linear versioning of the spec but in some other way, having the spec version communicated explicitly would help transitioning to that new feature detection approach. E.g., if version is 1.5.0+, use XYZ mechanism of feature detection. This should allow more freedom in defining feature detection mechanism (e.g., either in the base spec or in the extension), no need to decide on it now.

> However, I think that a media type parameter specifying a json-api-version is too rigid because it implies that media type negotiation will involve the version. 

I think it implies that because the question of versioning has been intentionally left to be ambiguous. If versioning schema and use of related headers are explicitly defined in the spec, nothing will have to be implied anymore.

> It also implies that extensions themselves should be versioned, which leads to an n-dimensional grid of complexity.

IMHO, if n-dimensional complexity arises, the reason would be not absence/presence of version in headers but the fact that extensions are allowed to develop independently. Versioning just adds some formalization coating to the existing development model, it does not define it. If the spec just says that versions apply to the spec as a whole, including official extensions (which can be changed in the future), nothing is implied anymore.

> So, what to do about this? If it's really the only issue with using a media type parameter (which I kinda doubt)

An alternative approach could be to place version in a separate mandatory header:

```
X-JSON-API-version: 1.0.0
```

It may be a better approach, but I wonder whether a separate header mandated by exchange format, not the application itself, could hurt adoption of the format.

It can also be recommended for custom extensions to define their versions as parameters of this header.

**Comment by ethanresnick** [comment]
It sounds like we all agree on having versions at least for documentation purposes, so I can create a PR to add a changelog-style page once we hit 1.0.

As far as communicating the version with each request: if that's something we decide to do, I can't imagine us making it mandatory. I.e. rather than failing requests that don't include a version, I think we'd say that clients/servers should process them in accordance with whatever iteration of the base spec they happen to know about—since the whole point is that we're guaranteeing that all json-api documents will be backwards/forwards compatible.

Therefore, if providing versioning information is going to be optional, there's no reason we have to figure it out now; we can finally launch 1.0 and then have a more rigorous discussion about different approaches later (including about ones not discussed yet, like simply having a key in the payload).

The only thing that I do think we should figure out pre-1.0 is whether @taion's suggestion to move a couple of the already-optional pieces of the base spec into extensions makes sense. I see a lot of potential value there, and I'm going to open another issue about it later today.

**Comment by taion** [comment]
I agree with @ethanresnick's above comment in its entirety.

**Comment by hhware** [comment]
> As far as communicating the version with each request: if that's something we decide to do, I can't imagine us making it mandatory

If there is no interest in making it mandatory, perhaps an official extension can be created, which offers a way to communicate it? Since the only purpose and benefit of explicit communication of version is to simplify interoperability, I still think it is important to have a way to do so which is _endorsed by the spec_ rather than let implementations to offer custom ways to do it, and an official extension is a straightforward way to do it. I also still believe it is important to do it before 1.0 for the exact same reason: the day 1.0 is out, the term "legacy" starts to acquire meaning in relation to JSON API.

I think that in this case, a separate header would be the best option:

```
X-JSON-API-version: 1.0.0
```

It would be a trivial extension, I can prepare a PR for it very quickly.

Also, what about the part (1) of the initial posting? Nobody seems to have anything against it. I think, it is just a clear and concise way to communicate intentions and to reinforce the point about keeping the spec backward compatible.

**Comment by kellysutton** [comment]
If a specified version of JSON API is omitted by the client, I think the server should just assume v1.0.x. If new features are added, my gut says it should be up to the client to explicitly request that version (preferably through a media type, but a separate header works too).

If a client tries to use one of these features in a 1.x release, where x > 0, we can respond with a familiar status code and payload. It's also possible that the same status and payload that might be used for unsupported/not implemented extensions.

To me, this seems like a bit of an upgrade vs. fallback question. For things like APIs, I've found an upgrade-style approach with pithy errors works best: "To use the JSON API super-pagination feature, you need to explicitly request 1.1.0 in your media type." The error messages should never be seen by end-users, only developers working with the API.

**Comment by taion** [comment]
@kellysutton 

I don't understand the benefit of that proposal. Given that backward compatibility is being maintained, who benefits from the client getting his "super-pagination" request denied just because he forgot to include the appropriate header? I think the main effect would be to make it harder to write a client that fully complies with the spec and can use the new feature.

Realistically, I think the most common use case ought to be a developer writing to an API following documentation (most likely without any particular client library), with the benefit of JSON API making different APIs more consistent. I think the main avenue of communication of available functionality is going to be in API documentation, not via automatic discovery, in any event.

I can picture cases like a very friendly client library e.g. allowing the user to make filtered requests, but then doing the filtering on the client side if the server does not advertise support for filtering, but that actually seems more like a recipe for trouble than anything else.

I really like the idea of making programmatic version/feature discovery optional. I can see use cases for it, but I think the most common use cases will not need it, so it'd be good to make the minimal compliant server and client not need to have that functionality.

**Comment by kellysutton** [comment]
@taion 

You're right. The full backwards compatibility always trips me up a bit. I don't know the answer, but what happens when a client tries to use a 1.x feature against a 1.0 server? Distinguishing between those two cases by explicit versioning could be worth exploring.

I also agree on the documentation issue, but whenever possible, I advocate for having a spec that realizes that good documentation might not always exist.

That smart clients scenario sounds like a bit of a rat's next. You're right.

The minimal compliant server/client (talking JSON API 1.0) should not need to have this functionality. In that way, I think the "how to version" discussion could be tabled until post-1.0 as long as we acknowledge that future clients/server will need a way to specify the JSON API version.

**Comment by hhware** [comment]
> Realistically, I think the most common use case ought to be a developer writing to an API following documentation (most likely without any particular client library), with the benefit of JSON API making different APIs more consistent. I think the main avenue of communication of available functionality is going to be in API documentation, not via automatic discovery, in any event.

@taion, I agree that for the use case of "a developer writing to an API following documentation (most likely without any particular client library)", the ability to communicate JSON API version in request/response would rarely help because that API likely has its own versions, which can be tied to versions of JSON API being used via documentation of that particular API. However, in case of a generic library or a hypermedia API, it could be useful to have such ability.

**Comment by dgeb** [comment]
The base spec simply can't evolve to make demands upon servers and clients that aren't present in 1.0. If we introduce a backwards-incompatible MUST, for instance, then the spec will no longer be backwards compatible.

Of course, we can still introduce additive features that MAY be supported by a server. Clients won't know whether they're supported in a 1.1 server OR a 1.0 server, which makes them backwards compatible.

The bottom line is that, if we're going to make demands upon a client or server that aren't present in 1.0, those demands must be placed in an extension that can be negotiated so that both client and server agree upon them. We just can't make additional demands in the base spec, since it is negotiated based on the media type alone. And that's why we're taking 1.0 so seriously now.

**Comment by taion** [comment]
@kellysutton @hhware 

I think the "explicit versioning" extension that @hhware talked about earlier is a really good way to accomplish both providing a good standard for doing generic hypermedia stuff. I do think this adds a lot of value in sanity checking, but I don't think it's something that every JSON API compliant implementation should be required to offer.

@dgeb 

If JSON API can only add optional features, wouldn't it make more sense for those just to be "official" extensions, in some sense? Given that you've already specced out `supported-ext` and `ext` media type parameters, it would seem more parsimonious to define such additional optional features as just being extensions, since then you'd already have a way of negotiating support.

**Comment by dgeb** [comment]
> If JSON API can only add optional features, wouldn't it make more sense for those just to be "official" extensions, in some sense?

I see nothing wrong with having some MAYs in the base spec. This allows a response to include optional elements without requiring an official extension. If every MAY in the base spec was eliminated, then a base implementation couldn't include `self` links, pagination links, `related` links, `meta` data, etc. and would require clients to negotiate for inclusion of any these. This seems unnecessarily strict and limiting by default.

On the other hand, I see nothing wrong with extensions that essentially convert MAYs in the base spec into MUSTs. I've brought up the concept of a "hypermedia" extension which requires `self` and `related` links.

I am interested to see extensions evolve. I'm sure that some "unofficial" extensions will be officially endorsed in time.

**Comment by dgeb** [comment]
To reiterate: I propose that we tag versions of the spec on github and display a history of changes since 1.0. However, I don't think that the spec's version needs to be represented in the document structure, in headers, nor should it be part of media type negotiations. If we somehow mess up royally with 1.0, we'll need to mint a new media type down the road.

@steveklabnik @tkellen @wycats what do you think?

**Comment by taion** [comment]
@dgeb 

I take back what I just said. Of course you are right with respect to this. The part that strikes me as being a little odd are the statements where the base spec says that I **MAY** do something, but **MUST** do so in a particular way if I choose to do so.

One example is the sorting specification - it seems like to explicitly advertise that I support server-side sorting, I would need to have an extension that just says "this implementation **MUST** support sorting as defined in the base specification", which seems odd to me.

**Comment by dgeb** [comment]
@taion As we transitioned from RC1 to RC2, we converted a lot of SHOULDs to the MAY/MUST pattern. While SHOULDs lead to endless fragmentation, MAY/MUST pairs significantly limit branches for implementations to consider.

> One example is the sorting specification - it seems like to explicitly advertise that I support server-side sorting, I would need to have an extension that just says "this implementation MUST support sorting as defined in the base specification", which seems odd to me.

This pattern guarantees consistency in base implementations. It also allows non-extended implementations to support optional capabilities. 

---

Edit: please ignore the following, as corrected by @bintoro below:

~~Consider the following requirement in the Extending section:~~

> ~~Servers MUST NOT provide extended functionality that is incompatible with the base specification to clients that do not request the extension in the ext parameter of the Content-Type or the Accept header.~~

~~This means that a base implementation simply couldn't provide sorting capabilities, for instance, if sorting was only specified in an extension. This also applies to pagination links and other options.~~

**Comment by bintoro** [comment]
@dgeb When I wrote the clause

> Servers MUST NOT provide extended functionality that is incompatible with the base specification to clients that do not request the extension in the ext parameter [...]

I intended "incompatible" to be taken literally.

Let's imagine sorting was defined in an extension. That's purely additive and thus not incompatible. It would be fine for a server to advertise the feature in `supported-ext`, but I don't want to have to add `ext=sort` just to be able to sort on a field. The `sort` query string parameter is enough to indicate that I do, in fact, require some sorting.

The server can even enable the extension without _any_ trigger from the client (default sort).

**Comment by dgeb** [comment]
@bintoro thanks for the correction - you are of course correct. I struck out my statement above to limit confusion, and will try to avoid posting when I'm dead tired :/

**Comment by hhware** [comment]
@dgeb [wrote](https://github.com/json-api/json-api/pull/458#issuecomment-87399972):

> We are not versioning the official extensions along with the base spec.

Is it mentioned anywhere? To me, it is just another argument in favor of explicitly clarifying versioning strategy of JSON API _and_ its official extensions.

**Comment by DracoBlue** [comment]
@dgeb 

> To reiterate: I propose that we tag versions of the spec on github and display a history of changes since 1.0. However, I don't think that the spec's version needs to be represented in the document structure, in headers, nor should it be part of media type negotiations. If we somehow mess up royally with 1.0, we'll need to mint a new media type down the road.

:+1: Having the version number in tags on the repository is sufficient. If you want to display the latest version somewhere, one could use http://badge.fury.io/ or a similar service. This keeps version increments when merging pull-requests easy.

**Comment by hhware** [comment]
>  If we somehow mess up royally with 1.0, we'll need to mint a new media type down the road.

An interesting approach! 

If Russian roulette is [played by a pro](http://en.wikipedia.org/wiki/Russian_roulette), chances to win are much higher than `1/X where X in (5,6,7)`. But if the player is inaccurate or runs out of luck, he has to retire from the business. No doubt, in case of JSON API media type, `X` is much higher than 7, but to me, the similarity is creepy :)

**Comment by tkellen** [comment]
Resolved in #687!

**Comment by taion** [comment]
:clap:

**Comment by hhware** [comment]
> Resolved in #687!

@tkellen, I do not think #687 fully resolves this issue: it does define the way to communicate version X.Y to clients, but it does not define meaning of X and Y.

Why not adopt semver? Versions will always be 1.X.Y, which clearly communicates the notion of backward compatibility.

**Comment by tkellen** [comment]
The only meaning we can convey with a version number alone is that some new additive features _might_ be supported. It doesn't matter what versioning scheme we use. In all cases, the spec will be backwards compatible.

If you're looking for more meaning than that, it's possible to give each optional feature a name and include a manifest in `meta` under the `jsonapi` object that indicates which features are available. HTML has been wildly successful for 22 years without anything like this, though, so I'm not convinced it's needed. That said, we may someday support something like this normatively.
